ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/FileListStreamer.java

```diff
@@ -0,0 +1,137 @@
+  private volatile boolean valid = true;
+  private volatile boolean asyncMode = false;
+  private final Object COMPLETION_LOCK = new Object();
+  private volatile boolean completed = false;
+
+
+
+  public FileListStreamer(LinkedBlockingQueue<String> cache, Path backingFile, Configuration conf) throws IOException {
+    this.cache = cache;
+    this.backingFile = backingFile;
+    this.conf = conf;
+    init();
+  }
+
+  private void init() throws IOException {
+    FileSystem fs = FileSystem.get(backingFile.toUri(), conf);
+    backingFileWriter = new BufferedWriter(new OutputStreamWriter(fs.create(backingFile, !asyncMode)));
+    LOG.info("Initialized a file based store to save a list at: {}, ayncMode:{}", backingFile, asyncMode);
+  }
+
+  public boolean isValid() {
+    return valid;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (!asyncMode) {
+      closeBackingFile();
+      return;
+    }
+    stop = true;
+    synchronized (COMPLETION_LOCK) {
+      while (!completed && isValid()) {
+        try {
+          COMPLETION_LOCK.wait(TimeUnit.SECONDS.toMillis(TIMEOUT_IN_SECS));
+        } catch (InterruptedException e) {
+        }
+      }
+    }
+    if (!isValid()) {
+      throw new IOException("File list is not in a valid state:" + backingFile);
+    }
+    LOG.info("Completed close for File List backed by ", backingFile);
+  }
+
+  public synchronized void writeInThread(String nextEntry) throws SemanticException {
+    try {
+      backingFileWriter.write(nextEntry);
+      backingFileWriter.newLine();
+    } catch (IOException e) {
+      throw new SemanticException(e);
+    }
+  }
+  @Override
+  public void run() {
+    asyncMode = true;
+    boolean exThrown = false;
+    while (!exThrown && (!stop || !cache.isEmpty())) {
+      try {
+        String nextEntry = cache.poll(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
+        if (nextEntry != null) {
+          backingFileWriter.write(nextEntry);
+          backingFileWriter.newLine();
+          LOG.debug("Writing entry {} to file list backed by {}", nextEntry, backingFile);
+        }
+      } catch (Exception iEx) {
+        if (!(iEx instanceof InterruptedException)) {
+          // not draining any more. Inform the producer to avoid OOM.
+          valid = false;
+          LOG.error("Exception while saving the list to file " + backingFile, iEx);
+          exThrown = true;
+        }
+      }
+    }
+    try{
+      closeBackingFile();
+      completed = true;
+    } finally {
+      synchronized (COMPLETION_LOCK) {
+        COMPLETION_LOCK.notify();
+      }
+    }
+    LOG.info("Completed the file list streamer backed by: {}", backingFile);
+  }
+
+  private void closeBackingFile() {
+    try {
+      backingFileWriter.close();
+      LOG.debug("Closed the file list backing file: {}", backingFile);
+    } catch (IOException e) {
+      LOG.error("Exception while closing the file list backing file", e);
+      valid = false;
```

Comment:
you can directly throw the exception from here instead of maintaining another state in valid boolean

Reply:
No, both are in different threads. the producer will never come to know the state of consumer thread. valid sort of communicates that.